### PR TITLE
chore(ui): migrate Search Input component to Typescript

### PR DESCRIPTION
**Pull request by guoda-puidokaite**
# Summary

- Migrated the `Search Input` component to `TypeScript`, enhancing code quality, maintainability, and scalability.
- Utilised static typing to minimize runtime errors and improve the development experience.
- Enhanced unit tests, covering more edge cases.
- Improved documentation with clearer descriptions of the component and its props.

# Changes Made

**TypeScript Migration**:

- Converted `Search Input` component files to `TypeScript`.
- Added comprehensive type annotations.

**Refactoring**:

- Made necessary adjustments to address TypeScript-detected issues.
- Further refactored for cleaner and more understandable code e.g. changed function names to begin with a verb.

**Testing**:

- Added more Edge Cases:
  - Various Prop Values: Validated handling of different `names`, `placeholders`, and initial values.
  - Placeholder Handling: Checked both default and custom `placeholder` values.
  - Value Handling: Ensured specific `values` render correctly and `Clear` icon appears when input has `value`.
  - Special Characters and Enter Key: Verified behaviour of special characters and the `Enter` key. Updated the `onKeyPress` test and removed the `TODO`.
  - Attributes and ClassNames: Validated default and custom `names`, and ensured all props are applied.
  - Disabled State: Confirmed the component reflects the `disabled` state properly.
  - Clear Button: Verified appearance and function of the `Clear` button, including event handling.
- Organised tests by functionality for better readability and easier maintenance.
- Verified component behaviour using `Storybook`.
- Updated tests to use `vitest`.

**Deprecated Code:**

- Moved the deprecated `.js` component to `/deprecated_js` to be used in other component test files
- Removed deprecated `.js` component story

**Documentation:**

- Updated `Storybook` documentation.

Screenshot: Deprecated `.js` code on the left, migrated `.ts` code on the right (**see notes**)

Screenshot 2024-10-17 at 10 16 15

**Release:**

- Generated a `changeset` file, mentioning the migration and bug fix.

# Related Issues

- https://github.com/cloudoperators/juno/issues/247

# Testing Instructions

1. Run `Storybook`
2. Run `Search Input` tests

# Checklist

- [x] I have performed a self-review of my code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have made corresponding changes to the documentation (if applicable).
- [x] My changes generate no new warnings or errors.

### CODE REVIEW

**Comment by guoda-puidokaite** [comment]
Resolves issue https://github.com/cloudoperators/juno/issues/247

File: packages/ui-components/src/components/SearchInput/SearchInput.component.tsx [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+   */
+  onClick?: (_event: MouseEvent<HTMLButtonElement | HTMLAnchorElement>) => void
+  /**
+   * Change handler for the search input.
+   */
+  onChange?: (_event: ChangeEvent<HTMLInputElement>) => void
+  /**
+   * KeyPress handler for the search input. By default, triggers the onSearch function when the 'Enter' key is pressed.
+   */
+  onKeyPress?: (_event: KeyboardEvent<HTMLInputElement>) => void
```
Comment (guoda-puidokaite) [inline packages/ui-components/src/components/SearchInput/SearchInput.component.tsx:66]: FYI that `onKeyPress` is deprecated https://github.com/cloudoperators/juno/issues/247#issuecomment-2418881656

File: packages/ui-components/src/components/SearchInput/SearchInput.component.tsx [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+   */
+  disabled?: boolean
+  /**
+   * Custom placeholder text displayed in the search input.
+   */
+  placeholder?: string
+  /**
+   * Initial value for the search input. Note that this value will not be updated if changed by the user.
+   */
+  value?: string
```
Comment (guoda-puidokaite) [inline packages/ui-components/src/components/SearchInput/SearchInput.component.tsx:36]: This line in the description "Note that this value will not be updated if changed by the user." is a bit confusing. If we set a value and the user types something else, it does change. Or do we mean that it's not changed permanently? Maybe we can remove?
Screenshot 2024-10-17 at 10 17 27

File: packages/ui-components/src/components/SearchInput/SearchInput.component.tsx [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    if (event.key === "Enter" && onSearch) onSearch(val)
+    onKeyPress && onKeyPress(event)
+  }
+
+  const handleSearchClick: React.MouseEventHandler<HTMLButtonElement | HTMLAnchorElement> = (event) => {
+    onSearch && onSearch(val)
+    onClick && onClick(event)
+  }
+
+  const handleClearClick: React.MouseEventHandler<HTMLButtonElement | HTMLAnchorElement> = (event) => {
```
Comment (taymoor89) [inline packages/ui-components/src/components/SearchInput/SearchInput.component.tsx:190]: I'd suggest to wrap all event handlers `handleInputChange`, ` handleKeyPress`, `handleSearchClick` and `handleClearClick` inside `useCallback` hook otherwise react will have to recreate those methods on each re-render.

**Comment by guoda-puidokaite** [comment]
Hi @taymoor89, Thank You for the review! 💪 I didn't refactor any of the existing code to keep within the requirements of the ticket. But we can do it. 🚀 I added the suggested changes and some more of my own. You can re-review.

**Review by taymoor89 — APPROVED** [review]

File: packages/ui-components/src/components/SearchInput/SearchInput.component.tsx (hunk 2 above)
Comment (edda) [reply packages/ui-components/src/components/SearchInput/SearchInput.component.tsx:36]: yes you're right, this is outdated. What this was referring to was the fact that at some point in the lifetime of this component any updates to this value after the component was rendered wouldn't change the internal value of the input. But as you said, this is no longer the case because we ensure internal state update via the `useEffect` hook. So we could change the description as follows:

"Set the input value"

**Review by edda — CHANGES_REQUESTED** [review]
There is a rogue `0` in the searchinput next to the search icon from somewhere (see screenshot)
Screenshot 2024-10-17 at 15 34 44

File: packages/ui-components/src/components/SearchInput/SearchInput.component.tsx [hunk 4]
```diff
@@ -207,7 +207,7 @@ export const SearchInput: React.FC<SearchInputProps> = ({
           {...props}
         />
         <div className={getIconWrapperStyles(variant)}>
-          {clear && val?.length && (
+          {clear && val?.length ? (
```
Comment (taymoor89) [inline packages/ui-components/src/components/SearchInput/SearchInput.component.tsx:210]: normally it is better to be more explicit when checking for length for example if you do it like `val?.length > 0` then you can keep your refactoring rather than converting it back to conditional operator.

```clear && val?.length > 0 && (```

Comment (taymoor89) [reply packages/ui-components/src/components/SearchInput/SearchInput.component.tsx:210]: or you can also cast it to boolean like `!!val?.length`

Comment (guoda-puidokaite) [reply packages/ui-components/src/components/SearchInput/SearchInput.component.tsx:210]: Sure, that works too

**Review by edda — APPROVED** [review]

**Review by taymoor89 — APPROVED** [review]
